### DISCUSSION TITLE
Discussion: contribution process

### DISCUSSION THREAD

**Discussion by Tao-VanJS** [vanjs-org/van#37] [General]
Want to discuss the draft contribution process to see if there are any problems with 3rd party contributors. Once the steps can be finalized and verified by some contributors, I will publish them to `CONTRIBUTING.md`.

1. Install dependencies in the root directories:
```
npm install
```
2. Upgrade the version in the package.json file. Considering the significance of this feature, it might be suitable to increment the minor version.
3. Modify source files in the `src/` directory.
4. Execute `bash publish.sh` within the `src/` directory. (While it might seem unusual to run a command within a subdirectory, fixing that isn't currently a priority.) Make sure the bash script completes with no error messages. This action generates the published bundles corresponding to the new version.
5. Perform a browser-based test. Essentially, you need a static HTTP server to host the root directory. Then, attempt to load http://localhost:{port}/test/van.test.html in your browser to see if all tests pass.
6. If applicable, update the corresponding `.debug.js` and `.d.ts` files as well.
7. Develop new tests to encompass the new APIs.
8. Ideally, providing sample web pages to showcase the new feature would be beneficial.

**Comment by enpitsuLin** [comment]
IMO, In most JavaScript opensource community, people should prefer just contribute code/doc, publish by project maintainer

And end products(van-latest.js/van-latest.min.js) also shouldn't be contained in source code repo, which can be distribute by npm or just in Github release page

**Reply by enpitsuLin** [reply]
if distribute vanjs to npm, the version/publish should be no more one step in the contribution process

**Reply by Tao-VanJS** [reply]
Sure. You can commit only the source file in a custom branch. I will do the publishing and merge it to the main :-) This is my first open source project thus I'm in the process of learning.

**Reply by andrewgryan** [reply]
I agree with @enpitsuLin the release process should be decoupled from the contribution process, to give the maintainers a chance to stabilise the release or combine multiple pull requests into a single release. I’ll revert the version bump in my PR to trial the above approach

**Comment by FredericHeem** [comment]
Would it be possible to add prettier to ensure a common styling ?

**Reply by Tao-VanJS** [reply]
It should be possible. Might not be the current priority of VanJS project. I can take a look at this when I have a big chunk of time.

**Comment by Tao-VanJS** [comment]
Thanks for @enpitsuLin and @andrewgryan for your precious feedback. I think the steps described above can be used for testing purpose, if you want to (you can also skip this part if your PR is simple enough and you have the confidence in its correctness).

For 3rd-party PRs, I think the process can be:
1. Contributor submits a PR that only contains changes on `van.js`, with `van.debug.js`, `van.d.ts` or `van.test.ts` if applicable.
2. VanJS maintainer will merge the PR in a custom branch.
3. VanJS maintainer will be responsible for the version upgrade, completing the testing, and then merging the custom branch to main.

Let me know your thought on this.

**Reply by FredericHeem** [reply]
Only `van.test.ts` should be edited, and `van.test.js` should be in `.gitignore` as it is a generated file.

**Reply by Tao-VanJS** [reply]
`van.test.js` is the test file for `van.js`.

**Reply by FredericHeem** [reply]
Correct me if i am wrong but any modification to `van.test.js` will be lost when running `build.sh`

**Reply by Tao-VanJS** [reply]
Oh, my bad. What I meant is `van.test.ts`. Edited the original comment.

**Reply by andrewgryan** [reply]
I'm curious, what is the purpose of `van.debug.js`?

**Reply by Tao-VanJS** [reply]
https://vanjs.org/start#download-table has the description of the debug file.
